### PR TITLE
Support native parquet writer in hive module and Misc fixes

### PR DESCRIPTION
**Pull request by qqibrow**

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveModule.java [hunk 1]
```diff
@@ -27,6 +27,7 @@
 import io.prestosql.plugin.hive.orc.OrcPageSourceFactory;
 import io.prestosql.plugin.hive.orc.OrcReaderConfig;
 import io.prestosql.plugin.hive.orc.OrcWriterConfig;
+import io.prestosql.plugin.hive.parquet.ParquetFileWriterFactory;
```
Comment (dain) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveModule.java:30]: The commit message has a typo `Add paruqet writer in hive module`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/parquet/ParquetFileWriter.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        catch (Exception e) {
+            throw new PrestoException(HIVE_WRITER_CLOSE_ERROR, "Error rolling back write parquet to Hive", e);
+        }
+    }
+
+    @Override
+    public long getValidationCpuNanos()
+    {
+        return 0;
+    }
```
Comment (dain) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/parquet/ParquetFileWriter.java:150]: Consider adding a `toString` for debugging.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/parquet/ParquetFileWriterFactory.java [hunk 3]
```diff
@@ -18,6 +18,7 @@
 import io.prestosql.plugin.hive.HdfsEnvironment;
 import io.prestosql.plugin.hive.HiveConfig;
 import io.prestosql.plugin.hive.HiveFileWriterFactory;
+import io.prestosql.plugin.hive.HiveSessionProperties;
```
Comment (dain) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/parquet/ParquetFileWriterFactory.java:21]: This should be in the previous commit

File: presto-hive/src/main/java/io/prestosql/plugin/hive/parquet/ParquetWriterConfig.java [hunk 4]
```diff
@@ -47,6 +50,19 @@ public ParquetWriterConfig setPageSize(DataSize pageSize)
         return this;
     }
 
+    public boolean isParquetOptimizedWriterEnabled()
+    {
+        return parquetOptimizedWriterEnabled;
+    }
+
+    @Config("hive.parquet.optimized-writer.enabled")
+    @ConfigDescription("Enable optimized parquet writer")
```
Comment (dain) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/parquet/ParquetWriterConfig.java:59]: Capitalize `Parquet` in description

File: presto-parquet/pom.xml [hunk 5]
```diff
@@ -82,6 +82,12 @@
             <artifactId>slice</artifactId>
         </dependency>
 
+        <dependency>
```
Comment (dain) [inline presto-parquet/pom.xml:85]: This should be in the commit `Support getRetainedBytes in ParquetWriter`

File: presto-parquet/pom.xml [hunk 6]
```diff
@@ -82,6 +82,12 @@
...
+            <groupId>org.openjdk.jol</groupId>
+            <artifactId>jol-core</artifactId>
+            <scope>provided</scope>
```
Comment (dain) [inline presto-parquet/pom.xml:88]: I don't think this scope is correct.  In the ORC module it not `provided`.

File: presto-parquet/src/main/java/io/prestosql/parquet/writer/PrimitiveColumnWriter.java [hunk 7]
```diff
@@ -296,7 +296,10 @@ private void flushCurrentPageToBuffer()
     @Override
     public long getBufferedBytes()
     {
-        return pageBuffer.stream().mapToLong(ParquetDataOutput::size).sum() + definitionLevelEncoder.getBufferedSize() + repetitionLevelEncoder.getBufferedSize();
+        return pageBuffer.stream().mapToLong(ParquetDataOutput::size).sum() +
```
Comment (dain) [inline presto-parquet/src/main/java/io/prestosql/parquet/writer/PrimitiveColumnWriter.java:303]: I would squash this into the commit that added getBufferedBytes.  I think it was `Expose written bytes and buffered bytes in ParquetWriter `

File: presto-parquet/src/main/java/io/prestosql/parquet/writer/PrimitiveColumnWriter.java [hunk 8]
```diff
@@ -102,6 +103,8 @@ public PrimitiveColumnWriter(Type type, ColumnDescriptor columnDescriptor, Primi
         this.compressionCodec = requireNonNull(compressionCodecName, "compressionCodecName is null");
         this.compressor = getCompressor(compressionCodecName);
         this.pageSizeThreshold = pageSizeThreshold;
+
+        this.columnStatistics = Statistics.createStats(columnDescriptor.getPrimitiveType());
```
Comment (dain) [inline presto-parquet/src/main/java/io/prestosql/parquet/writer/PrimitiveColumnWriter.java:107]: Are there statistics for non-primitive types?

File: presto-parquet/src/main/java/io/prestosql/parquet/writer/ParquetWriter.java [hunk 9]
```diff
@@ -85,6 +90,16 @@ public ParquetWriter(OutputStream outputStream, List<String> columnNames, List<T
...
     }
 
+    public long getWrittenBytes()
+    {
+        return outputStream.size();
+    }
+
+    public long getBufferedBytes()
+    {
+        return columnWriters.stream().mapToLong(ColumnWriter::getBufferedBytes).sum();
```
[the remark this answers is not in the document]
Comment (qqibrow) [reply presto-parquet/src/main/java/io/prestosql/parquet/writer/ParquetWriter.java:113]: I resolved this. Also, I don't quite understand why in OrcWriter:
```
    @Override
    public long getWrittenBytes()
    {
        return orcWriter.getWrittenBytes() + orcWriter.getBufferedBytes();
    }

```
why bufferedBytes is part of written bytes?

File: presto-parquet/src/main/java/io/prestosql/parquet/writer/PrimitiveColumnWriter.java (hunk 8 above)
Comment (qqibrow) [reply presto-parquet/src/main/java/io/prestosql/parquet/writer/PrimitiveColumnWriter.java:107]: No. This aims to fill the stats part for primitive columns https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L740 . There is no support for this on non-primitive types.

**Pull request merged**